### ISSUE TITLE
Incompatible Override Self Distinction

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#5288]
**Describe the bug**
```python
from __future__ import annotations

from typing import Any, Generic, TypeVar, overload

T = TypeVar("T")


class Foo(Generic[T]):
    @overload
    def transform(self: Foo[int], x: list[int]) -> list[int]:
        ...

    @overload
    def transform(self, x: Any) -> Any:
        ...

    def transform(self, x: Any) -> Any:
        ...


class Bar(Foo[str]):
    def transform(self, x: str) -> dict[str, str]:
        ...
```

Produces a type error of incompatible override.

**Expected behavior**
I would expect no error. The Foo[int] overload should not be applicable because it's only if T = int, while Bar has T set to str.

**VS Code extension or command-line**
pyright 1.1.314

**Additional context**
The real version I hit with this error can actually be found with this [definition](https://github.com/python/typeshed/blob/8624edd3a5e4b2b20208bc55a425a62ff6585691/stubs/tensorflow/tensorflow/keras/layers.pyi#L65) in typeshed. One overload handles most common case as most of the time typevar has that specific type. Other overload fallsback to Any as I don't think it's possible to describe behavior in general today. That approach seems to produce incompatible overrides now.

I'll also note mypy --strict passes here and doesn't mark the override as incompatible. Unsure if it has self special rules or is just laxer here.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.315, which I just published. It will also be included in this week's insiders release of pylance.
